### PR TITLE
jwt updates

### PR DESCRIPTION
**Pull request by damencho**
Moves all the jwt token verification logic inside token/util so it can be reused.
Adds the logic about room name and domain name verification to toke/util.
Adds some room_size API changes, that uses the token/util for room and domain verification and jwt verification. Also, changes mod_room_size from global to host specific module. 
Everything is split in different commits but I can extract the room_size changes in separate commit if considered to be easier.

### CODE REVIEW

File: resources/prosody-plugins/token/util.lib.lua [hunk 1]
```diff
@@ -1,11 +1,108 @@
...
+
+    if asapKeyServer and not have_async then
+        module:log("error", "requires a version of Prosody with util.async");
+        return true;
+    end
+
+    return false;
+end
+
+local http_timeout = 30;
```
Comment (paweldomas) [inline resources/prosody-plugins/token/util.lib.lua]: I know it was like that before, but could you please move "http_timeout" and "http_headers" vars declarations to the top, next to the others ?

File: resources/prosody-plugins/token/util.lib.lua [hunk 2]
```diff
@@ -30,11 +127,62 @@ local function _verify_token(token, appId, appSecret, disableRoomNameConstraints
 		return nil, "'room' claim is missing";
 	end
 
+    local audClaim = claims["aud"];
```
Comment (paweldomas) [inline resources/prosody-plugins/token/util.lib.lua]: in future please do not mix up adding new logic with moving things around especially if the commit msg says "this commit moves things to.."

File: resources/prosody-plugins/token/util.lib.lua [hunk 3]
```diff
@@ -1,11 +1,108 @@
...
+local path = require "util.paths";
+local sha256 = require "util.hashes".sha256;
+local timer = require "util.timer";
+
 local jwt = require "luajwtjitsi";
 
 local _M = {};
 
-local function _verify_token(token, appId, appSecret, disableRoomNameConstraints)
+local appId = module:get_option_string("app_id");
```
Comment (paweldomas) [inline resources/prosody-plugins/token/util.lib.lua]: After thinking about it some more I find it strange that utility module is loading config values. It's hard to tell which module owns this config options and where this values should be declared.

File: resources/prosody-plugins/mod_token_verification.lua [hunk 4]
```diff
@@ -50,25 +52,17 @@ local function verify_user(session, stanza)
...
-		log("error",
-			"Unable to get name of the MUC room ? to: %s", stanza.attr.to);
-		return nil;
-	end
-
 	local token = session.auth_token;
 	local auth_room = session.jitsi_meet_room;
-	if disableRoomNameConstraints ~= true and room ~= string.lower(auth_room) then
+	local auth_domain = session.jitsi_meet_domain;
+	if disableRoomNameConstraints ~= true and token_util.verify_room_and_domain(stanza.attr.to, auth_room, auth_domain) ~= true then
```
Comment (paweldomas) [inline resources/prosody-plugins/mod_token_verification.lua]: I think that token_util may not be getting right config values for "mod_token_verification" which is supposed to load them for the parent host. Or at least it's hard to figure out just by looking at the code.

File: resources/prosody-plugins/token/util.lib.lua [hunk 5]
```diff
@@ -177,6 +185,39 @@ local function _verify_token(session, token)
...
+            log("error",
+                "Unable to get name of the MUC room ? to: %s", room_address_to_verify);
+            return false;
+        end
+        return room_to_verify == string.lower(room);
+    end
+
+    -- make sure we remove any resource when comparing
+    room_address_to_verify = jid.bare(room_address_to_verify);
+    local target_subdomain, target_room = room_to_verify:match("^%[([^%]]+)%](.+)$");
```
Comment (paweldomas) [inline resources/prosody-plugins/token/util.lib.lua]: what is the format of this multidomain thing ? could you please add a comment with example ?

File: resources/prosody-plugins/token/util.lib.lua [hunk 6]
```diff
@@ -177,6 +185,39 @@ local function _verify_token(session, token)
...
+
+        -- from this point we depend on muc_domain_base, deny access if option is missing
+        if not muc_domain_base then
+            module:log("warn", "No 'muc_domain_base' option set, denying access!");
+            return false;
+        end
+
+        return room_address_to_verify == jid.join("["..domain.."]"..string.lower(room), muc_domain);
+    else
+        -- we do not have a domain part (multidomain is not enabled)
```
Comment (paweldomas) [inline resources/prosody-plugins/token/util.lib.lua]: does the config option 'enableDomainVerification' mean "multidomain is enabled" ?

Comment (paweldomas) [inline resources/prosody-plugins/token/util.lib.lua]: is there any doc available anywhere that explain this options with some examples ? it seems like there are bunch of possible scenarios, but I'm not able to verify this code without any docs

File: resources/prosody-plugins/mod_muc_size.lua [hunk 7]
```diff
@@ -39,6 +40,33 @@ function get_room_from_jid(room_jid)
 	end
 end
 
+function verify_token(token, room_address)
```
Comment (paweldomas) [inline resources/prosody-plugins/mod_muc_size.lua]: don't we have similar code in other modules ?

File: resources/prosody-plugins/token/util.lib.lua [hunk 8]
```diff
@@ -30,11 +135,99 @@ local function _verify_token(token, appId, appSecret, disableRoomNameConstraints
...
+    -- now verify the whole token
+    local claims, msg;
+    if asapKeyServer then
+        claims, msg = _verify_jwt_token(token, appId, pubKey, disableRoomNameConstraints);
+    else
+        claims, msg = _verify_jwt_token(token, appId, appSecret, disableRoomNameConstraints);
+    end
+    if claims ~= nil then
+        -- Binds room name to the session which is later checked on MUC join
+        session.jitsi_meet_room = claims["room"];
```
Comment (paweldomas) [inline resources/prosody-plugins/token/util.lib.lua]: It feels weird that method named "_verify_token" which takes 'session' as an argument magically adds some attributes to it. It should be documented with some warning at least in the method description. Or maybe the name should be different.

Also I have the impression that all methods have grown significantly and there is no documentation whatsoever to any config options, so now it's quite hard to figure out what's going on.

**Comment by damencho** [comment]
Closing this one. As working on a new approach where util is a class and can be instantiated.
New one is https://github.com/jitsi/jitsi-meet/pull/1548

**Pull request closed without merging**